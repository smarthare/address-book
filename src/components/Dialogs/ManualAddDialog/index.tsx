import { useContext, useEffect, useState } from "react";

import DialogBase from "../DialogBase";
import { BequestInput, BequestAutoSelect } from "components/bequest";

import { AddressContext } from "contexts/AddressContext";
import countries from "constants/countires";
import { DialogEnhancedProps } from "types/props";
import { Address } from "types/address";
import { StyledButton } from "./style";

const ManualAddDialog = (props: DialogEnhancedProps) => {
  const { handleOpenNext, handleClose, ...rest } = props;

  const { editId, addrs, setAddress, updateAddress } =
    useContext(AddressContext);

  const [addr, setAddr] = useState<Address>();
  const [country, setCountry] = useState<string>();
  const [line1, setLine1] = useState<string>();
  const [line2, setLine2] = useState<string>();
  const [line3, setLine3] = useState<string>();
  const [town, setTown] = useState<string>();
  const [county, setCounty] = useState<string>();
  const [postcode, setPostcode] = useState<string>();

  useEffect(() => {
    if (addr) {
      setLine1(addr.line_1);
      setLine2(addr.line_2);
      setLine3(addr.line_3);
      setTown(addr.town_or_city);
      setCounty(addr.county);
      setCountry(addr.country);
      setPostcode(addr.postcode);
    }
  }, [addr]);

  useEffect(() => {
    if (editId) {
      const _addr = addrs.find((addr) => addr.id === editId);
      setAddr(_addr);
    } else {
      setAddr(undefined);
    }
  }, [editId, addrs]);

  const handleSave = () => {
    if (line1 && town && country && postcode) {
      const newAddr = {
        line_1: line1,
        line_2: line2,
        line_3: line3,
        town_or_city: town,
        county,
        country,
        formatted_address: [line1, line2, line3, town, county],
      };

      editId
        ? updateAddress(editId, newAddr, postcode)
        : setAddress(newAddr, postcode);
      handleClose();
    }
  };

  return (
    <DialogBase
      handleOkAction={handleSave}
      handleCancelAction={handleClose}
      handleClose={handleClose}
      {...rest}
    >
      <StyledButton onClick={handleOpenNext} data-testid="dialogAutoBtn">
        Lookup address by postcode
      </StyledButton>

      <BequestInput
        placeholder="Line 1"
        key="line_1"
        value={addr && addr.line_1}
        handleChange={setLine1}
        gutterTop
        required
      />
      <br />

      <BequestInput
        placeholder="Line 2"
        key="line_2"
        value={addr && addr.line_2}
        handleChange={setLine2}
        gutterTop
      />
      <br />

      <BequestInput
        placeholder="Line 3"
        value={addr && addr.line_3}
        handleChange={setLine3}
        gutterTop
      />
      <br />

      <BequestInput
        placeholder="Town"
        value={addr && addr.town_or_city}
        handleChange={setTown}
        required
        gutterTop
      />
      <br />

      <BequestInput
        placeholder="County"
        value={addr && addr.county}
        handleChange={setCounty}
        gutterTop
      />
      <br />

      <BequestAutoSelect
        placeholder="Select Country"
        options={countries}
        defaultValue={addr ? addr.country : null}
        handleChange={setCountry}
      />
      <br />

      <BequestInput
        placeholder="Postcode"
        value={addr && addr.postcode}
        handleChange={setPostcode}
        gutterTop
        required
      />
      <br />
    </DialogBase>
  );
};

export default ManualAddDialog;
