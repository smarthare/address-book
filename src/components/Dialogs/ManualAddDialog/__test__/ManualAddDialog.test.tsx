import { render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";

import ManualAddDialog from "../../ManualAddDialog";

afterEach(() => {
  cleanup();
});

test("Should render ManualAddDialog component", () => {
  render(
    <ManualAddDialog
      open={true}
      handleClose={() => {}}
      handleOpenNext={() => {}}
    />
  );

  const dialogElement = screen.getByRole("dialog");

  expect(dialogElement).toBeInTheDocument();
  expect(dialogElement).toHaveClass("MuiPaper-root");
  expect(screen.getByTestId("dialogAutoBtn")).toHaveTextContent(
    "Lookup address by postcode"
  );
  expect(screen.getByTestId("dialogContent").childElementCount).toEqual(
    6 + 4 + 10 + 1 + 1
  );
});

test("ManualAddDialog matches snapshot", () => {
  const component = render(
    <ManualAddDialog
      open={true}
      handleClose={() => {}}
      handleOpenNext={() => {}}
    />
  );

  expect(component).toMatchSnapshot();
});
