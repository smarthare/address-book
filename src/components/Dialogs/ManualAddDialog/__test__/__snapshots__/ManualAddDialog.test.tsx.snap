// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManualAddDialog matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root MuiModal-root css-liyq5o-MuiModal-root-MuiDialog-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby="customized-dialog-title"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXl MuiDialog-paperFullWidth css-csmqbv-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-17guxnh-MuiTypography-root-MuiDialogTitle-root"
            id="customized-dialog-title"
          >
            <div
              class="css-ed9sft"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="HomeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                />
              </svg>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1tvtw5b-MuiTypography-root"
              data-testid="dialogTitle"
            />
            <button
              aria-label="close"
              class="MuiButtonBase-root css-1kv8631-MuiButtonBase-root"
              tabindex="0"
              type="button"
            >
              <div
                class="css-1qcze82"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CloseIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </h2>
          <div
            class="MuiDialogContent-root MuiDialogContent-dividers css-1t4vnk2-MuiDialogContent-root"
            data-testid="dialogContent"
          >
            <button
              class="css-1tv4pea"
              data-testid="dialogAutoBtn"
            >
              Lookup address by postcode
            </button>
            <input
              class="css-1m9nqg5"
              placeholder="Line 1"
              value=""
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-lpk256-MuiTypography-root"
            >
              Required
            </p>
            <br />
            <br />
            <input
              class="css-1m9nqg5"
              placeholder="Line 2"
              value=""
            />
            <br />
            <input
              class="css-1m9nqg5"
              placeholder="Line 3"
              value=""
            />
            <br />
            <input
              class="css-1m9nqg5"
              placeholder="Town"
              value=""
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-lpk256-MuiTypography-root"
            >
              Required
            </p>
            <br />
            <br />
            <input
              class="css-1m9nqg5"
              placeholder="County"
              value=""
            />
            <br />
            <div
              class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-dd2h8b-MuiAutocomplete-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1ahezau-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-autocomplete="list"
                    aria-expanded="false"
                    aria-invalid="false"
                    autocapitalize="none"
                    autocomplete="off"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id="controllable-states-demo"
                    placeholder="Select Country"
                    role="combobox"
                    spellcheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Open"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      tabindex="-1"
                      title="Open"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-hdw1oc"
                    >
                      <span
                        class="notranslate"
                      >
                        â€‹
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-lpk256-MuiTypography-root"
            >
              Required
            </p>
            <br />
            <input
              class="css-1m9nqg5"
              placeholder="Postcode"
              value=""
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-lpk256-MuiTypography-root"
            >
              Required
            </p>
            <br />
            <br />
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root css-zaye0n-MuiButtonBase-root"
              tabindex="0"
              type="button"
            >
              Cancel
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root css-1f1k8ev-MuiButtonBase-root"
              tabindex="0"
              type="button"
            >
              OK
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
